vyper/semantics/types/user.py

```diff
@@ -281,9 +282,58 @@ def from_EventDef(cls, base_node: vy_ast.EventDef) -> "EventT":
         return cls(base_node.name, members, indexed, base_node)
 
     def _ctor_call_return(self, node: vy_ast.Call) -> None:
-        validate_call_args(node, len(self.arguments))
-        for arg, expected in zip(node.args, self.arguments.values()):
-            validate_expected_type(arg, expected)
+        # Handle positional args by converting them to kwargs
+        # TODO: Remove block when positional args are removed
+        if len(node.args) > 0:
+            if len(node.keywords) > 0:
+                # can't mix args and kwargs
+                raise InstantiationException(
+                    "Event instantiation requires either all positional arguments "
+                    "or all keyword arguments",
+                    node,
+                )
+
+            msg = "Instantiating events with positional arguments is "
+            msg += "deprecated as of v0.4.1 and will be disallowed "
+            msg += "in a future release. Use kwargs instead eg. "
+            msg += "Foo(a=1, b=2)"
+
+            vyper_warn(msg, node)
+
+            # convert positional args to keywords
+            kw_list = []
+            for kw, val in zip(self.arguments.keys(), node.args):
+                kw_node = vy_ast.keyword(arg=kw, value=val)
+                kw_node.set_parent(node)
+                kw_list.append(kw_node)
+            node.keywords = kw_list
+            node.args = []
+
+        # manually validate kwargs for better error messages instead of
+        # relying on `validate_call_args` (same as structs)
+        members = self.arguments.copy()
+        keys = list(self.arguments.keys())
+        for i, kwarg in enumerate(node.keywords):
+            # x=5 => kwarg(arg="x", value=Int(5))
+            argname = kwarg.arg
+            if argname not in members:
+                hint = get_levenshtein_error_suggestions(argname, members, 1.0)
+                raise UnknownAttribute("Unknown or duplicate event argument.", kwarg, hint=hint)
+            expected = keys[i]
+            if argname != expected:
+                raise InvalidAttribute(
```

Comment:
not covered by tests (https://github.com/vyperlang/vyper/pull/4257/checks?check_run_id=30669010870)

Reply:
added test covering these cases